openfisca_core/parameters.py

```diff
@@ -559,17 +561,178 @@ def __getattr__(self, key):
+        MESSAGE_PART_3 = "To use fancy indexing on parameter node, its children must be homogenous."
+        MESSAGE_PART_4 = "See more at https://doc.openfisca.org/xxx."
+
+        def raise_key_inhomogeneity_error(node_with_key, node_without_key, missing_key):
+            message = " ".join([
+                MESSAGE_PART_1,
+                "'{}' exists, but '{}' doesn't.",
+                MESSAGE_PART_3,
+                MESSAGE_PART_4,
+                ]).format(
+                node._name,
+                '.'.join([node_with_key, missing_key]),
+                '.'.join([node_without_key, missing_key]),
+                ).encode('utf-8')
+
+            raise ValueError(message)
+
+        def raise_type_inhomogeneity_error(node_name, non_node_name):
+            message = " ".join([
+                MESSAGE_PART_1,
+                "'{}' is a node, but '{}' is not.",
+                MESSAGE_PART_3,
+                MESSAGE_PART_4,
+                ]).format(
+                node._name,
+                node_name,
+                non_node_name,
+                ).encode('utf-8')
+
+            raise ValueError(message)
+
+        def raise_not_implemented(node_name, node_type):
+            message = " ".join([
+                MESSAGE_PART_1,
+                "'{}' is a '{}', and fancy indexing has not been implemented yet on this kind of parameters.",
+                MESSAGE_PART_4,
+                ]).format(
+                node._name,
+                node_name,
+                node_type,
+                ).encode('utf-8')
+            raise NotImplementedError(message)
+
+        def extract_named_children(node):
+            return {
+                '.'.join([node._name, key]): value
+                for key, value in node._children.iteritems()
+                }
+
+        def check_nodes_homogeneous(named_nodes):
+            """
+                Check than several nodes (or parameters, or baremes) have the same structure.
+            """
+            names = named_nodes.keys()
+            nodes = named_nodes.values()
+            first_node = nodes[0]
+            first_name = names[0]
+            if isinstance(first_node, ParameterNodeAtInstant):
+                children = extract_named_children(first_node)
+                for node, name in zip(nodes, names)[1:]:
+                    if not isinstance(node, ParameterNodeAtInstant):
+                        raise_type_inhomogeneity_error(first_name, name)
+                    first_node_keys = first_node._children.keys()
+                    node_keys = node._children.keys()
+                    if not first_node_keys == node_keys:
+                        missing_keys = set(first_node_keys).difference(node_keys)
+                        if missing_keys:
+                            raise_key_inhomogeneity_error(first_name, name, missing_keys.pop())
+                        missing_key = set(node_keys).difference(first_node_keys).pop()
```

Comment:
Would it be helpful to explicitly put an `else`here ?

Reply:
In this case, why not.
Added a comment too.